virtualizarr/manifests/indexing.py

```diff
@@ -0,0 +1,176 @@
+from zarr.core.indexing import (
+    BasicIndexer,
+    BasicSelection,
+    BasicSelector,
+    IntDimIndexer,
+    SliceDimIndexer,
+)
+
+
+# TODO write a custom message for this
+class SubChunkIndexingError(IndexError): ...
+
+
+def array_indexer_to_chunk_grid_indexer(
+    indexer: BasicIndexer,
+    arr_shape: tuple[int, ...],
+    chunk_shape: tuple[int, ...],
+) -> BasicIndexer:
+    """
+    Translate an indexer into an array into a corresponding indexer into the underlying chunk grid.
+
+    As compressed chunks cannot be subdivided, this will raise on any array slices that require slicing within individual chunks.
+
+    Parameters
+    ----------
+    indexer
+    arr_shape : tuple[int, ...]
+        Shape of the ManifestArray we are trying to index into.
+    """
+
+    print(f"{indexer=}")
+
+    [
+        print(chunk_coords, chunk_selection, out_selection, is_complete_chunk)
+        for chunk_coords, chunk_selection, out_selection, is_complete_chunk in indexer
+    ]
+
+    # TODO move this check outside? Because we can arbitrarily index into uncompressed arrays...
+    # TODO or is that unrelated? Uncompressed arrays allow us to choose arbitrary chunking at reference generation time
+    if not all(is_complete_chunk for _, _, _, is_complete_chunk in indexer):
+        raise SubChunkIndexingError()
+
+    if indexer.drop_axes:
+        raise NotImplementedError
```

Comment:
This never triggers, and I don't know why. Either a bug in `zarr.core.indexing` or I don't understand this part of `zarr.core.indexing` (more likely).